File: quant1x/contrib/data/tdx/level1/std/hello.rs
// -*- coding: utf-8 -*-
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// hello — 标准行情握手1/2 (STD_SYNCHRONIZE1, STD_SYNCHRONIZE2)

use encoding_rs::GBK;

use super::super::super::command::*;
use super::super::super::protocol::{BaseFrame, RequestHeader, ResponseHeader};

// ============================================================
// StdLoginContext (STD_SYNCHRONIZE1, 0x000d) — 标准行情握手1
// ============================================================

#[derive(Debug, Clone)]
pub struct StdLoginContext {
    pub req_header: RequestHeader,
    pub resp_header: ResponseHeader,
    pub info: String,
}

impl BaseFrame for StdLoginContext {
    fn request_header(&self) -> &RequestHeader { &self.req_header }
    fn request_header_mut(&mut self) -> &mut RequestHeader { &mut self.req_header }
    fn response_header(&self) -> &ResponseHeader { &self.resp_header }
    fn response_header_mut(&mut self) -> &mut ResponseHeader { &mut self.resp_header }

    fn serialize_request_body(&mut self) -> Vec<u8> {
        let padding = hex::decode("01").unwrap_or_default();
        padding
    }

    fn deserialize_response_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        let offset = 68usize;
        if data.len() >= offset {
            let info_bytes = &data[offset..];
            let (cow, _, _) = GBK.decode(info_bytes);
            self.info = cow.into_owned();
        }
        Ok(())
    }
}

impl StdLoginContext {
    pub fn new() -> Self {
        Self {
            req_header: RequestHeader::new(STD_SYNCHRONIZE1, FLAG_UNCOMPRESSED),
            resp_header: ResponseHeader::new(),
            info: String::new(),
        }
    }
}

// ============================================================
// UpgradeTipContext (STD_SYNCHRONIZE2, 0x0fdb) — 标准行情握手2
// ============================================================

#[derive(Debug, Clone)]
pub struct UpgradeTipContext {
    pub req_header: RequestHeader,
    pub resp_header: ResponseHeader,
    pub info: String,
}

impl BaseFrame for UpgradeTipContext {
    fn request_header(&self) -> &RequestHeader { &self.req_header }
    fn request_header_mut(&mut self) -> &mut RequestHeader { &mut self.req_header }
    fn response_header(&self) -> &ResponseHeader { &self.resp_header }
    fn response_header_mut(&mut self) -> &mut ResponseHeader { &mut self.resp_header }

    fn serialize_request_body(&mut self) -> Vec<u8> {
        hex::decode("d5d0c9ccd6a4a8af0000008fc22540130000d500c9ccbdf0d7ea00000002")
            .unwrap_or_default()
    }

    fn deserialize_response_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        let offset = 58usize;
        if data.len() >= offset {
            let (cow, _, _) = GBK.decode(&data[offset..]);
            self.info = cow.into_owned();
        }
        Ok(())
    }
}

impl UpgradeTipContext {
    pub fn new() -> Self {
        Self {
            req_header: RequestHeader::new(STD_SYNCHRONIZE2, FLAG_UNCOMPRESSED),
            resp_header: ResponseHeader::new(),
            info: String::new(),
        }
    }
}

File: quant1x/contrib/data/tdx/level1/std/hello.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import struct
from datetime import datetime

from ...command import Command
from ... import protocol


class StdLoginContext(protocol.BaseFrame):
    """第一次协议握手: login"""
    def __init__(self):
        super().__init__(Command.STD_SYNCHRONIZE1)
        self.info = ""
        self._padding = bytes.fromhex("01")

    def serialize_request_body(self) -> bytes:
        return self._padding

    def deserialize_response_body(self, data: bytes) -> None:
        (
            _, year, day, month, minute, hour, _, second,
            unknown1, unknown2, unknown3,
            date, a1, b1, date2, a2, b2,
            unknown4, unknown5, unknown6,
            server_name, web_site, unknown7, category,
        ) = struct.unpack('<BHBBBBBB16s16sBIHHIHHHH5s22s64s6s30s', data)

        info = {
            "date_time": datetime(year, month, day, hour, minute, second).strftime('%Y-%m-%d %H:%M:%S'),
            "server_name": server_name.decode('gbk').replace('\x00', ''),
            "web_site": web_site.decode('gbk').replace('\x00', ''),
            "category": category.decode('gbk').replace('\x00', ''),
            "b1": b1, "b2": b2, "a1": a1, "a2": a2,
            "date": date, "date2": date2,
            "unknown1": unknown1, "unknown2": unknown2, "unknown3": unknown3,
            "unknown4": unknown4, "unknown5": unknown5, "unknown6": unknown6,
            "unknown7": unknown7,
        }
        self.reply = info


class UpgradeTipContext(protocol.BaseFrame):
    """第二次协议握手"""
    def __init__(self):
        super().__init__(Command.STD_SYNCHRONIZE2)
        self.info = ""
        self._padding = bytes.fromhex("d5d0c9ccd6a4a8af0000008fc22540130000d500c9ccbdf0d7ea00000002")

    def serialize_request_body(self) -> bytes:
        return self._padding

    def deserialize_response_body(self, data: bytes) -> None:
        offset = 58
        if len(data) >= offset:
            info_bytes = data[offset:]
            try:
                self.info = info_bytes.decode('gbk', errors='ignore').rstrip('\x00')
            except Exception:
                self.info = info_bytes.decode('utf-8', errors='ignore')
